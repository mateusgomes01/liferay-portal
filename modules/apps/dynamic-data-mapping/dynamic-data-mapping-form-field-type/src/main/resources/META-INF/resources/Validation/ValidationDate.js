/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayInput, ClaySelectWithOption} from '@clayui/form';
import {PagesVisitor, useFormState} from 'data-engine-js-components-web';
import React, {useMemo, useState} from 'react';

import Select from '../Select/Select.es';
import StartEndDate from './StartEndDate';

import './ValidationDate.scss';
import {EVENT_TYPES} from './validationReducer';

const customDateLabel = Liferay.Language.get('custom-date');
const endsOnLabel = Liferay.Language.get('ends-on');
const responseDateLabel = Liferay.Language.get('response-date');
const startsFromLabel = Liferay.Language.get('starts-from');

const responseDateOption = {
	checked: false,
	label: responseDateLabel,
	name: 'responseDate',
	value: 'responseDate',
};

const getDateOptionsByType = (label, name) => ({
	label,
	name,
	options: [
		responseDateOption,
		{
			checked: false,
			label: customDateLabel,
			name: 'customDate',
			value: 'customDate',
		},
	],
});

const getOperation = (quantity) => {
	return quantity < 0 ? 'minus' : 'plus';
};

const getSelectedParameter = (
	value,
	selectedParameterName,
) => {
	if (value && typeof value === 'string') {
		try {
			value = JSON.parse(value);
		}
		catch (error) {}
	}

	return value?.[selectedParameterName];
};

const getSignedValue = (operation, value) => {
	let signedValue = value;

	if (
		(operation === 'minus' && value > 0) ||
		(operation === 'plus' && value < 0)
	) {
		signedValue = value * -1;
	}

	return signedValue;
};

const parameters = {
	dateRange: [
		getDateOptionsByType(startsFromLabel, 'startsFrom'),
		getDateOptionsByType(endsOnLabel, 'endsOn'),
	],
	futureDates: [getDateOptionsByType(startsFromLabel, 'startsFrom')],
	pastDates: [getDateOptionsByType(endsOnLabel, 'endsOn')],
};

const ValidationDate = ({
	dispatch,
	errorMessage,
	localizationMode,
	localizedValue,
	name,
	onBlur,
	parameter,
	parentFieldName,
	readOnly,
	selectedValidation,
	transformSelectedValidation,
	validations,
	visible,
}) => {

	const startDate = getSelectedParameter(
		localizedValue(parameter),
		'startsFrom',
	);
	const endDate = getSelectedParameter(
		localizedValue(parameter),
		'endsOn',
	);
	const selectedParameter = parameters[selectedValidation.name];
	
	const handleChangeParameters = (typeName, parameters) => {

		const parameter = {}

		if (typeName === 'startsFrom') {
			parameter.startsFrom = parameters;
			parameter.endsOn = endDate;
			
		} else if (typeName === 'endsOn') {
			parameter.endsOn = parameters;
			parameter.startsFrom = startDate;
		}

		dispatch({payload: {parameter}, type: EVENT_TYPES.SET_PARAMETER});
	};

	const errorMessageName = name + '_errorMessage';
	const getDateTypeValue = (name) => {
		return name === 'startsFrom' ? startsFrom : endsOn;
	};

	const {builderPages} = useFormState();

	
	const fields = useMemo(() => {
		const fields = [];
		const visitor = new PagesVisitor(builderPages);

		visitor.visitFields((field) => {
			if (
				!field.repeatable &&
				field.type === 'date' &&
				field.fieldName !== parentFieldName
			) {
				fields.push({
					checked: false,
					label: field.label,
					name: field.fieldName,
					value: field.fieldName,
				});
			}
		});

		return fields;
	}, [builderPages]);

	return (
		<>
			<label>
				{Liferay.Language.get('accepted-date')}
				<ClaySelectWithOption 
					disabled={readOnly || localizationMode}
					name="selectedValidation"
					onChange={({target: {value}}) => {
						dispatch({
							payload: {
								selectedValidation: transformSelectedValidation(
									value
								),
							},
							type: EVENT_TYPES.CHANGE_SELECTED_VALIDATION,
						});
					}}
					options={validations}
					value={selectedValidation.name}		
				/>
			</label>

			{selectedParameter.map(({label, name: eventType, options}, index) => {

				const {title, tooltip, parameters} = eventType === 'startsFrom' ? {
					title: Liferay.Language.get('start-date'),
					tooltip: Liferay.Language.get(
						'starts-from-tooltip'
				  ),
				  	parameters: startDate
				} : {
					title: Liferay.Language.get('end-date'),
					tooltip: Liferay.Language.get('ends-on-tooltip'),
					parameters: endDate
				}

				return (
					<React.Fragment key={index}>
						{selectedParameter.length > 1 && (
							<>
								<label>{title.toUpperCase()}</label>
								<div className="separator" />
							</>
						)}

						<StartEndDate
							dateFieldOptions={fields}
							eventType={eventType}
							label={label}
							name={name}
							onChange={handleChangeParameters}
							options={options}
							parameters={parameters}
							readOnly={localizationMode || readOnly }
							tooltip={tooltip}
							visible={visible}
						/>
					</React.Fragment>
				);
			})}
			<label htmlFor={errorMessageName}>
				{Liferay.Language.get('error-message')}
			</label>
			<ClayInput
				id={errorMessageName}
				name={errorMessageName}
				onBlur={onBlur}
				onChange={(event) => {
					dispatch({
						payload: {
							errorMessage: event.target.value,
						},
						type: EVENT_TYPES.CHANGE_ERROR_MESSAGE,
					});
				}}
				placeholder={Liferay.Language.get('this-field-is-invalid')}
				type="text"
				value={localizedValue(errorMessage)}
			/>
		</>
	);
};

export default ValidationDate;
