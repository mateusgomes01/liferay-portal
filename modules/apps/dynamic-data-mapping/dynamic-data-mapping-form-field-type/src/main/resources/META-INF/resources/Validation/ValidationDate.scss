.ddm-form-field-type__validation-date {
	width: 33%;
}

.ddm__validation-date-accepted-date {
	margin-bottom: 24px;
}
