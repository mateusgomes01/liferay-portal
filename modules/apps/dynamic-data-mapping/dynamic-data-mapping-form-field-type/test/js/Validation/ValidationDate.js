/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import '@testing-library/jest-dom/extend-expect';
import {cleanup, render} from '@testing-library/react';
import {FormProvider} from 'data-engine-js-components-web';
import React from 'react';

import ValidationDate from '../../../src/main/resources/META-INF/resources/Validation/ValidationDate';

const globalLanguageDirection = Liferay.Language.direction;

const validations = [
	{
		checked: false,
		label: 'Future Dates',
		name: 'futureDates',
		parameterMessage: '',
		template: 'futureDates({name}, "{parameter}")',
		value: 'futureDates',
	},
	{
		checked: false,
		label: 'Past Dates',
		name: 'pastDates',
		parameterMessage: '',
		template: 'pastDates({name}, "{parameter}")',
		value: 'pastDates',
	},
	{
		checked: false,
		label: 'Range',
		name: 'dateRange',
		parameterMessage: '',
		template: 'dateRange({name}, "{parameter}")',
		value: 'dateRange',
	},
];

const generateParameter = () => ({
	en_US: {
		endsOn: {
			date: 'responseDate',
			dateFieldName: 'Date1234',
			quantity: 1,
			type: 'customDate',
			unit: 'days',
		},
		startsFrom: {
			date: 'responseDate',
			dateFieldName: 'Date1234',
			quantity: 1,
			type: 'customDate',
			unit: 'days',
	},
}

})

const parameters = generateParameter();

const localizedValue = jest.fn(() => parameters['en_US']);

const ValidationDateProvider = ({builderPages = [], ...props}) => (
	<FormProvider initialState={{builderPages}}>
		<ValidationDate {...props} />
	</FormProvider>
);

describe('ValidationDate', () => {
	beforeAll(() => {
		Liferay.Language.direction = {
			en_US: 'rtl',
		};
	});

	afterAll(() => {
		Liferay.Language.direction = globalLanguageDirection;
	});

	afterEach(cleanup);

	it('shows future dates validation', () => {

		const {container} = render(
			<ValidationDateProvider
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				localizedValue={localizedValue}
				parameter={parameters}
				name="validationDate"
				onChange={() => {}}
				selectedValidation={{
					label: '',
					name: 'futureDates',
					parameterMessage: '',
					template: 'futureDates({name}, "{parameter}")',
				}}
				validations={validations}
				visible={true}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('shows past dates validation', () => {
		const {container} = render(
			<ValidationDateProvider
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				localizedValue={localizedValue}
				parameter={parameters}
				name="validationDate"
				onChange={() => {}}
				selectedValidation={{
					label: '',
					name: 'pastDates',
					parameterMessage: '',
					template: 'pastDates({name}, "{parameter}")',
				}}
				validations={validations}
				visible={true}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('shows date range validation', () => {
		const {container} = render(
			<ValidationDateProvider
				builderPages={[]}
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				localizedValue={localizedValue}
				parameter={parameters}
				name="validationDate"
				onChange={() => {}}
				selectedValidation={{
					label: '',
					name: 'dateRange',
					parameterMessage: '',
					template: 'dateRange({name}, "{parameter}")',
				}}
				validations={validations}
				visible={true}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('shows custom date fields for Future dates', () => {
		const parameter = {
			en_US: {
				startsFrom: {
					date: 'responseDate',
					quantity: 1,
					type: 'customDate',
					unit: 'days',
				},
			},
		};

		const localizedValue = jest.fn(() => parameter['en_US']);
		const {getAllByRole} = render(
			<ValidationDateProvider
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				localizedValue={localizedValue}
				name="validationDate"
				onChange={() => {}}
				parameter={parameter}
				selectedValidation={{
					label: '',
					name: 'futureDates',
					parameterMessage: '',
					template: 'futureDates({name}, "{parameter}")',
				}}
				validations={validations}
				visible={true}
			/>
		);

		const [selectedValidaiton, selectedOperation, selectedUnit] = [...getAllByRole('combobox')];

		const startQuantity = getAllByRole('textbox')

		expect(selectedValidaiton).toHaveValue('futureDates');
		expect(selectedOperation).toHaveValue('plus');
		expect(selectedUnit).toHaveValue('days');
		expect(startQuantity[0]).toHaveValue(1);

	});

	it('shows custom date fields for Past dates and operation minus when quantity is negative', () => {
		const parameter = {
			en_US: {
				endsOn: {
					date: 'responseDate',
					quantity: -1,
					type: 'customDate',
					unit: 'days',
				},
			},
		};

		const localizedValue = jest.fn(() => parameter['en_US']);
		const {getAllByRole} = render(
			<ValidationDateProvider
				defaultLanguageId="en_US"
				editingLanguageId="en_US"
				localizedValue={localizedValue}
				name="validationDate"
				onChange={() => {}}
				parameter={parameter}
				selectedValidation={{
					label: '',
					name: 'pastDates',
					parameterMessage: '',
					template: 'pastDates({name}, "{parameter}")',
				}}
				validations={validations}
				visible={true}
			/>
		);

		const [acceptedDate, operation, unit] = [...getAllByRole('listbox')];

		const [quantity] = [...getAllByRole('textbox')];

		expect(acceptedDate).toHaveValue('pastDates');
		expect(operation).toHaveValue('minus');
		expect(quantity).toHaveValue(1);
		expect(unit).toHaveValue('days');
	});
});
