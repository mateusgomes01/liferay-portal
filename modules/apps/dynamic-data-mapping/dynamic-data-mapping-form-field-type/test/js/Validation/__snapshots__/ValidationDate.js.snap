// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationDate shows date range validation 1`] = `
<div>
  <div
    class="ddm__validation-date-accepted-date"
  >
    <label>
      accepted-date
    </label>
    <select
      class="form-control"
      name="selectedValidation"
    >
      <option
        name="futureDates"
        parametermessage=""
        template="futureDates({name}, \\"{parameter}\\")"
        value="futureDates"
      >
        Future Dates
      </option>
      <option
        name="pastDates"
        parametermessage=""
        template="pastDates({name}, \\"{parameter}\\")"
        value="pastDates"
      >
        Past Dates
      </option>
      <option
        name="dateRange"
        parametermessage=""
        template="dateRange({name}, \\"{parameter}\\")"
        value="dateRange"
      >
        Range
      </option>
    </select>
  </div>
  <label>
    START-DATE
  </label>
  <div
    class="separator"
  />
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        starts-from
      </label>
      <div
        data-tooltip-align="top"
        title="starts-from-tooltip"
      >
        <svg
          class="lexicon-icon lexicon-icon-question-circle-full ddm__validation-date-start-end-icon"
          role="presentation"
        >
          <use
            xlink:href="#question-circle-full"
          />
        </svg>
      </div>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            custom-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        date
      </label>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            response-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm-custom-date"
  >
    <label>
      operation
      <select
        class="form-control"
        name="selectedOperation_startsFrom"
      >
        <option
          value="plus"
        >
          plus
        </option>
        <option
          value="minus"
        >
          minus
        </option>
      </select>
    </label>
    <label>
      quantity
      <input
        class="form-control ddm-field-text"
        max="999"
        min="1"
        name="inputedQuantity_startsFrom"
        type="number"
        value="1"
      />
    </label>
    <label>
      unit
      <select
        class="form-control"
        name="selectedUnit_startsFrom"
      >
        <option
          value="days"
        >
          days
        </option>
        <option
          value="months"
        >
          months
        </option>
        <option
          value="years"
        >
          years
        </option>
      </select>
    </label>
  </div>
  <label>
    END-DATE
  </label>
  <div
    class="separator"
  />
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        ends-on
      </label>
      <div
        data-tooltip-align="top"
        title="ends-on-tooltip"
      >
        <svg
          class="lexicon-icon lexicon-icon-question-circle-full ddm__validation-date-start-end-icon"
          role="presentation"
        >
          <use
            xlink:href="#question-circle-full"
          />
        </svg>
      </div>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            custom-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        date
      </label>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            response-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm-custom-date"
  >
    <label>
      operation
      <select
        class="form-control"
        name="selectedOperation_endsOn"
      >
        <option
          value="plus"
        >
          plus
        </option>
        <option
          value="minus"
        >
          minus
        </option>
      </select>
    </label>
    <label>
      quantity
      <input
        class="form-control ddm-field-text"
        max="999"
        min="1"
        name="inputedQuantity_endsOn"
        type="number"
        value="1"
      />
    </label>
    <label>
      unit
      <select
        class="form-control"
        name="selectedUnit_endsOn"
      >
        <option
          value="days"
        >
          days
        </option>
        <option
          value="months"
        >
          months
        </option>
        <option
          value="years"
        >
          years
        </option>
      </select>
    </label>
  </div>
  <label
    for="validationDate_errorMessage"
  >
    error-message
  </label>
  <input
    class="form-control"
    id="validationDate_errorMessage"
    name="validationDate_errorMessage"
    placeholder="this-field-is-invalid"
    type="text"
    value="[object Object]"
  />
</div>
`;

exports[`ValidationDate shows future dates validation 1`] = `
<div>
  <div
    class="ddm__validation-date-accepted-date"
  >
    <label>
      accepted-date
    </label>
    <select
      class="form-control"
      name="selectedValidation"
    >
      <option
        name="futureDates"
        parametermessage=""
        template="futureDates({name}, \\"{parameter}\\")"
        value="futureDates"
      >
        Future Dates
      </option>
      <option
        name="pastDates"
        parametermessage=""
        template="pastDates({name}, \\"{parameter}\\")"
        value="pastDates"
      >
        Past Dates
      </option>
      <option
        name="dateRange"
        parametermessage=""
        template="dateRange({name}, \\"{parameter}\\")"
        value="dateRange"
      >
        Range
      </option>
    </select>
  </div>
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        starts-from
      </label>
      <div
        data-tooltip-align="top"
        title="starts-from-tooltip"
      >
        <svg
          class="lexicon-icon lexicon-icon-question-circle-full ddm__validation-date-start-end-icon"
          role="presentation"
        >
          <use
            xlink:href="#question-circle-full"
          />
        </svg>
      </div>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            custom-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        date
      </label>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            response-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm-custom-date"
  >
    <label>
      operation
      <select
        class="form-control"
        name="selectedOperation_startsFrom"
      >
        <option
          value="plus"
        >
          plus
        </option>
        <option
          value="minus"
        >
          minus
        </option>
      </select>
    </label>
    <label>
      quantity
      <input
        class="form-control ddm-field-text"
        max="999"
        min="1"
        name="inputedQuantity_startsFrom"
        type="number"
        value="1"
      />
    </label>
    <label>
      unit
      <select
        class="form-control"
        name="selectedUnit_startsFrom"
      >
        <option
          value="days"
        >
          days
        </option>
        <option
          value="months"
        >
          months
        </option>
        <option
          value="years"
        >
          years
        </option>
      </select>
    </label>
  </div>
  <label
    for="validationDate_errorMessage"
  >
    error-message
  </label>
  <input
    class="form-control"
    id="validationDate_errorMessage"
    name="validationDate_errorMessage"
    placeholder="this-field-is-invalid"
    type="text"
    value="[object Object]"
  />
</div>
`;

exports[`ValidationDate shows past dates validation 1`] = `
<div>
  <div
    class="ddm__validation-date-accepted-date"
  >
    <label>
      accepted-date
    </label>
    <select
      class="form-control"
      name="selectedValidation"
    >
      <option
        name="futureDates"
        parametermessage=""
        template="futureDates({name}, \\"{parameter}\\")"
        value="futureDates"
      >
        Future Dates
      </option>
      <option
        name="pastDates"
        parametermessage=""
        template="pastDates({name}, \\"{parameter}\\")"
        value="pastDates"
      >
        Past Dates
      </option>
      <option
        name="dateRange"
        parametermessage=""
        template="dateRange({name}, \\"{parameter}\\")"
        value="dateRange"
      >
        Range
      </option>
    </select>
  </div>
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        ends-on
      </label>
      <div
        data-tooltip-align="top"
        title="ends-on-tooltip"
      >
        <svg
          class="lexicon-icon lexicon-icon-question-circle-full ddm__validation-date-start-end-icon"
          role="presentation"
        >
          <use
            xlink:href="#question-circle-full"
          />
        </svg>
      </div>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            custom-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm__validation-date-start-end"
  >
    <div
      class="ddm__validation-date-start-end-label"
    >
      <label>
        date
      </label>
    </div>
    <div
      class="dropdown"
    >
      <div
        class="dropdown-toggle form-builder-select-field input-group-container"
        style=""
      >
        <div
          class="form-control results-chosen select-field-trigger"
        >
          <div
            class="option-selected"
          >
            response-date
          </div>
          <a
            class="select-arrow-down-container"
          >
            <svg
              class="lexicon-icon lexicon-icon-caret-double"
              role="presentation"
            >
              <use
                xlink:href="#caret-double"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ddm-custom-date"
  >
    <label>
      operation
      <select
        class="form-control"
        name="selectedOperation_endsOn"
      >
        <option
          value="plus"
        >
          plus
        </option>
        <option
          value="minus"
        >
          minus
        </option>
      </select>
    </label>
    <label>
      quantity
      <input
        class="form-control ddm-field-text"
        max="999"
        min="1"
        name="inputedQuantity_endsOn"
        type="number"
        value="1"
      />
    </label>
    <label>
      unit
      <select
        class="form-control"
        name="selectedUnit_endsOn"
      >
        <option
          value="days"
        >
          days
        </option>
        <option
          value="months"
        >
          months
        </option>
        <option
          value="years"
        >
          years
        </option>
      </select>
    </label>
  </div>
  <label
    for="validationDate_errorMessage"
  >
    error-message
  </label>
  <input
    class="form-control"
    id="validationDate_errorMessage"
    name="validationDate_errorMessage"
    placeholder="this-field-is-invalid"
    type="text"
    value="[object Object]"
  />
</div>
`;
